File: src/Cocoar.SignalARRR.Common/HARRRError.cs
using System;
using System.Text.Json;

namespace Cocoar.SignalARRR.Common {
    /// <summary>
    /// Structured error envelope for SignalARRR exceptions.
    /// The server serializes it as pure JSON into the HubException message string.
    /// </summary>
    /// <remarks>
    /// <see cref="Code"/> is the machine-readable contract (see <see cref="HARRRErrorCodes"/>);
    /// <see cref="Message"/> is for humans; <see cref="Type"/> is the .NET exception type name and
    /// exists for .NET-side diagnostics only — non-.NET clients should never need it.
    /// </remarks>
    public class HARRRError {
        /// <summary>
        /// Version of this envelope's shape. 1 since the error contract rework; absent (0) on
        /// messages from older servers.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("Version")]
        public int Version { get; set; }

        /// <summary>
        /// Machine-readable error code — the field clients branch on. Unknown or missing codes
        /// must be treated as <see cref="HARRRErrorCodes.Internal"/>.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("Code")]
        public string? Code { get; set; }

        [System.Text.Json.Serialization.JsonPropertyName("Type")]
        public string Type { get; set; } = "Error";

        [System.Text.Json.Serialization.JsonPropertyName("Message")]
        public string Message { get; set; } = "";

        [System.Text.Json.Serialization.JsonPropertyName("StackTrace")]
        public string? StackTrace { get; set; }

        /// <summary>
        /// The cause chain, nested instead of flattened: previously only
        /// <c>GetBaseException()</c> survived the wire, which discarded every intermediate step.
        /// </summary>
        [System.Text.Json.Serialization.JsonPropertyName("InnerError")]
        public HARRRError? InnerError { get; set; }

        /// <summary>
        /// The <see cref="Code"/> folded to a value this client version knows
        /// (unknown/missing → <see cref="HARRRErrorCodes.Internal"/>).
        /// </summary>
        [System.Text.Json.Serialization.JsonIgnore]
        public string NormalizedCode => HARRRErrorCodes.Normalize(Code);

        /// <summary>
        /// Try to parse a HubException message into a structured HARRRError.
        /// Supports pure JSON (current servers), the SignalR-prefixed form, and the legacy
        /// [Type] Message format.
        /// </summary>
        public static HARRRError Parse(string message) {
            // Current servers put pure JSON into HubException.Message; on some paths SignalR
            // prefixes it ("... HARRRException: {json}"), so the JSON part is extracted first.
            var jsonCandidate = message;
            var marker = "HARRRException: ";
            var markerIndex = message.IndexOf(marker, StringComparison.Ordinal);
            if (markerIndex >= 0) {
                jsonCandidate = message.Substring(markerIndex + marker.Length);
            }

            try {
                var error = JsonSerializer.Deserialize<HARRRError>(jsonCandidate);
                // Accepted when it carries actual error content — a versioned envelope, a code, or
                // a concrete type. A bare "{}" (or unrelated JSON) falls through to the fallbacks.
                if (error != null && (error.Version >= 1 || !string.IsNullOrEmpty(error.Code) || (!string.IsNullOrEmpty(error.Type) && error.Type != "Error"))) {
                    return error;
                }
            } catch {
                // Not JSON — try legacy format
            }

            // Legacy format: [Type] Message (may also be after the SignalR prefix)
            var match = System.Text.RegularExpressions.Regex.Match(message, @"\[([\w.]+)\]\s*(.*)");
            if (match.Success) {
                return new HARRRError {
                    Type = match.Groups[1].Value,
                    Message = match.Groups[2].Value,
                };
            }

            // Fallback
            return new HARRRError {
                Type = "Error",
                Message = message,
            };
        }

        /// <summary>
        /// Try to parse a HubException into a structured HARRRError.
        /// </summary>
        public static HARRRError Parse(Exception exception) {
            return Parse(exception.Message);
        }
    }
}

File: src/Cocoar.SignalARRR.Kotlin/signalarrr/src/main/kotlin/dev/cocoar/signalarrr/HARRRError.kt
package dev.cocoar.signalarrr

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * The machine-readable error codes of the wire contract. Clients branch on these — a code is the
 * only cross-language signal, a .NET type name means nothing here. Application code may put its
 * own codes on the wire (`HARRRException("room_full", ...)` on the server); those travel verbatim.
 * A code this client does not know folds to [INTERNAL].
 */
public object HARRRErrorCodes {
    public const val UNAUTHORIZED: String = "unauthorized"
    public const val METHOD_NOT_FOUND: String = "method_not_found"
    public const val INVALID_ARGUMENT_COUNT: String = "invalid_argument_count"
    public const val ARGUMENT_BINDING_FAILED: String = "argument_binding_failed"
    public const val CANCELLED: String = "cancelled"
    public const val TIMEOUT: String = "timeout"
    public const val NO_CLIENT_RESPONDED: String = "no_client_responded"
    public const val UPLOAD_SLOT_LIMIT_REACHED: String = "upload_slot_limit_reached"
    public const val INTERNAL: String = "internal"

    private val known = setOf(
        UNAUTHORIZED, METHOD_NOT_FOUND, INVALID_ARGUMENT_COUNT, ARGUMENT_BINDING_FAILED,
        CANCELLED, TIMEOUT, NO_CLIENT_RESPONDED, UPLOAD_SLOT_LIMIT_REACHED, INTERNAL,
    )

    /** Maps a wire code to one this client knows; unknown or missing codes become [INTERNAL]. */
    public fun normalize(code: String?): String = if (code != null && code in known) code else INTERNAL
}

/**
 * The structured error envelope the server puts into a `HubException` message as JSON.
 *
 * [code] is what to branch on; [message] is for humans; [type] is the .NET exception type and
 * exists for .NET-side diagnostics; [stackTrace] is present only from DEBUG servers; [innerError]
 * nests the cause chain.
 */
@Serializable
public data class HARRRError(
    @SerialName("Version") val version: Int = 0,
    @SerialName("Code") val code: String? = null,
    @SerialName("Type") val type: String = "Error",
    @SerialName("Message") val message: String = "",
    @SerialName("StackTrace") val stackTrace: String? = null,
    @SerialName("InnerError") val innerError: HARRRError? = null,
) {
    /** [code] folded to a value this client version knows. */
    val normalizedCode: String get() = HARRRErrorCodes.normalize(code)

    public companion object {
        private const val MARKER = "HARRRException: "
        private val LEGACY = Regex("\\[([\\w.]+)]\\s*(.*)", RegexOption.DOT_MATCHES_ALL)
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        /**
         * Parses a `HubException` message. Accepts the pure JSON envelope, the SignalR-prefixed
         * form (`... HARRRException: {json}`), and the legacy `[Type] Message` format; anything
         * else becomes an error of type `Error` carrying the raw message.
         */
        public fun parse(message: String): HARRRError {
            val markerIndex = message.indexOf(MARKER)
            val candidate = if (markerIndex >= 0) message.substring(markerIndex + MARKER.length) else message

            runCatching { json.decodeFromString(serializer(), candidate.trim()) }.getOrNull()?.let { parsed ->
                // Accepted when it carries actual error content — a versioned envelope, a code, or a
                // concrete type. A bare "{}" (or unrelated JSON) falls through to the fallbacks.
                if (parsed.version >= 1 || !parsed.code.isNullOrEmpty() || (parsed.type.isNotEmpty() && parsed.type != "Error")) {
                    return parsed
                }
            }

            LEGACY.find(message)?.let { m ->
                return HARRRError(type = m.groupValues[1], message = m.groupValues[2])
            }

            return HARRRError(type = "Error", message = message)
        }

        public fun parse(throwable: Throwable): HARRRError = parse(throwable.message ?: throwable.toString())
    }
}

/**
 * A SignalARRR call failed on the server. [error] is the parsed envelope, [rawMessage] the string
 * the server sent. `message` reads `[code] message` for a structured error.
 */
public class HARRRException(public val error: HARRRError, public val rawMessage: String) :
    HubInvocationException(formatMessage(error, rawMessage)) {

    /** Shortcut for `error.normalizedCode`. */
    public val code: String get() = error.normalizedCode

    private companion object {
        fun formatMessage(error: HARRRError, raw: String): String =
            if (error.type == "Error" && error.code == null) raw else "[${error.code ?: error.type}] ${error.message}"
    }
}
